signer/xpi/xpi.go

```diff
@@ -142,20 +154,145 @@ func (s *PKCS7Signer) Config() signer.Configuration {
-func (s *PKCS7Signer) SignFile(input []byte, options interface{}) (signer.SignedFile, error) {
+// isCOSEAlgBad checks whether a COSE algorithm is recognized as a
+// valid IANA algorithm name, and supported and enabled in autograph
+func isCOSEAlgBad(algName string) (alg *cose.Algorithm, err error) {
+	alg, algErr := cose.GetAlgByName(algName)
+	if algErr != nil {
+		err = errors.Wrapf(algErr, "xpi: unrecognized algorithm COSE Signing algorithm")
+		return
+	}
+	if _, exists := supportedCOSEAlgNames[alg.Name]; exists == false {
+		err = fmt.Errorf("xpi: COSE algorithm %s is not supported", algName)
+		return
+	}
+	return alg, nil
+}
+
+// SignFile takes an unsigned zipped XPI file and returns a signed XPI file
+func (s *PKCS7Signer) SignFile(input []byte, options interface{}) (signedFile signer.SignedFile, err error) {
 	var (
-		signedFile []byte
+		pkcs7Manifest []byte
+		manifest []byte
+		metas = []Metafile{}
+		opt Options
+		coseSigAlgs []*cose.Algorithm // COSE algorithms to sign with
 	)
-	manifest, sigfile, err := makeJARManifests(input)
+
+	opt, err = GetOptions(options)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot get options")
+	}
+	for _, algName := range opt.COSEAlgorithms {
+		alg, err := isCOSEAlgBad(algName)
+		if err != nil {
+			return nil, err
+		}
+		coseSigAlgs = append(coseSigAlgs, alg)
+	}
+
+	manifest, err = makeJARManifest(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
+	}
+
+	if len(coseSigAlgs) > 0 {
+		cn := opt.ID
+		if s.EndEntityCN != "" {
+			cn = s.EndEntityCN
+		}
+		if cn == "" {
+			return nil, errors.New("xpi: missing common name")
+		}
+		tmp := cose.NewSignMessage(manifest)
+		msg := &tmp
+
+		var coseSigners []*cose.Signer
+
+		// Add list of DER encoded intermediate certificates as message key id
+		msg.Headers.Protected["kid"] = [][]byte{s.issuerCert.Raw[:]}
+		msg.Headers.Protected = cose.CompressHeaders(msg.Headers.Protected)
+
+		for _, alg := range coseSigAlgs {
+			// create a cert and key
+			eeDERCert, eeKey, err := s.MakeDEREndEntity(cn, getKeyOptionsForCOSEAlg(alg))
+			if err != nil {
+				return nil, err
+			}
+
+			// create a COSE.Signer
+			signer, err := cose.NewSigner(eeKey)
+			if err != nil {
+				return nil, errors.Wrap(err, "xpi: COSE signer creation failed")
+			}
+			coseSigners = append(coseSigners, signer)
+
+			// create a COSE Signature holder
+			sig := cose.NewSignature()
+			sig.Headers.Protected["alg"] = alg.Name
+			sig.Headers.Protected["kid"] = [][]byte{eeDERCert[:]}
+			msg.AddSignature(sig)
+		}
+
+		external := []byte("")
+		randReader := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+		err = msg.Sign(randReader, external, cose.SignOpts{
+			GetSigner: func(index int, signature cose.Signature) (cose.Signer, error) {
+				return *coseSigners[index], nil
+			},
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: COSE signing failed")
+		}
+		// don't include signature in payload
```

Comment:
is that because we use a detached signature? some more details in comments would be useful

Reply:
Yeah, it's null per the grand unified signing doc and https://github.com/franziskuskiefer/cose-rust/blob/b53d999253321c0b20dd22434f74047d933fd4c2/src/decoder.rs#L213